Support dropout for the fused attention when max seqlen <= 512

zlsh80826 opened the pull request:
This PR does the following two changes:

- Enables dropout support for both the C API and JAX transformer: It should be noticed that the PRNG is different between the native JAX and the fused attention.
  1. Unlike the native JAX, the fused attention custom calls need different seeds on different devices. To achieve this, the `dropout_rng` is [splitted](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.split.html) into `len(devices)` seeds. Each seed is distributed (by sharding) on each device.
  2. Use `FusedAttnOffsetManager` to maintain the offset to ensure the offset is not overlapped in different fused attention calls.
  3. The `rng_state` for `nvte_fused_attn_*` is composed of the seed from framework and the offset from `FusedAttnOffsetManager`.
  4. With (i) and (ii), the following things are guaranteed
    a. Seeds are different on different gpus
    b. Seeds and offsets are different on different TransformerLayers
    c. Seeds and offsets are different on the different iterations
    d. Offsets are the same on different gpus

- Change the default dropout shape: The original implementation of `core_attention` generates the dropout mask with the shape `(b, h, 1, s_kv)` ([code](https://github.com/NVIDIA/TransformerEngine/blob/v0.8.0.1/transformer_engine/jax/flax/transformer.py#L167)). However, this dropout shape is not aligned to other framworks/PAXML/cuDNN fused attention. In this PR, the dropout mask changes to be the same shape with the `attn_weights` without broadcasting.

Future work:
- Check if t5x needs the old dropout shape, if it is required. We need to extend the [hidden_dropout_dims](https://github.com/NVIDIA/TransformerEngine/blob/v0.8.0.1/transformer_engine/jax/flax/transformer.py#L686) to have the different attention dropout shape.

zlsh80826 commented:
/te-ci

zlsh80826 commented:
/te-ci

jeng1220 commented:
@nouiz , @cyanguwa , and @timmoon10 ,
The cuDNN dropout for fused attention is ready. Please help for review.

zlsh80826 commented:
/te-ci

nouiz commented:
Can you confirm the PR description is up to date?

nouiz commented on transformer_engine/jax/cpp_extensions.py:
```diff
@@ -1679,7 +1681,7 @@ def lowering(ctx, grad_outputs, softmax_outputs, *, scale_factor):
                                                          scale_factor)
 
-        return out # out is iterable already
+        return out    # out is iterable already
```
nitpik: Normally, it is 2 whitespace before the `#` symbol.

zlsh80826 replied:
We had a yapf style config that specifies to use 4 whitespaces, but it is not added in the qa script. We should add them into the qa script to have the align style check. cc @jeng1220

nouiz commented on transformer_engine/jax/cpp_extensions.py:
```diff
@@ -1979,6 +1981,23 @@ def scaled_upper_triang_masked_softmax_bwd(grad_outputs: jnp.ndarray, softmax_ou
+    # Jax default use threefry PRNG, which generates 2 u32
+    if rng_state.dtype != jnp.uint32:
+        warnings.warn(f"Requested {rng_state.dtype=} is not available, and will be "
+                      f"casted to dtype uint32.")
```
Maybe specify that JAX threefry PRNG doesn't have this issue?
So user know how to remove that warning.

zlsh80826 replied:
Adding the instructions for removing the issue in c8615cf3866fa0266d20739ef221b1ccdb5e45ea

zlsh80826 commented:
> Can you confirm the PR description is up to date?

Yes, it is up to date

nouiz commented on transformer_engine/jax/cpp_extensions.py:
```diff
@@ -1979,6 +1981,23 @@ def scaled_upper_triang_masked_softmax_bwd(grad_outputs: jnp.ndarray, softmax_ou
 
 
+def _check_rng_state(rng_state):
+    # Jax can't bind None, create a dummy tensor for None
```
Can you confirm how this work?
If end user don't specify an rng, but set a dropout rate, we will generate a static seed?
If the user end up creating many such layers, all the dropout seed will end up being the same.
Is that right?

I don't like that. What about having a global rng that we split everything we need one?
This way, it will still be deterministic, but each layers will have a different seed everytimes.

zlsh80826 replied:
The condition rng_state=None should only occur when the dropout rate is zero., I will add an assertion for this case. So, the dummy tensor will only be created when the dropout rate is zero. This creation is necessary for JAX custom call binding, as JAX cannot bind a None tensor. This dummy tensor will not be used in the underlying kernels since the dropout rate is zero.

zlsh80826 replied:
Enhance the rng_state checking in https://github.com/NVIDIA/TransformerEngine/commit/c8615cf3866fa0266d20739ef221b1ccdb5e45ea. I add the check that the `rng_state` must not be `None` if dropout is enabled

nouiz commented on transformer_engine/jax/cpp_extensions.py:
```diff
@@ -2162,8 +2183,18 @@ def lowering(ctx, qkv, softmax_aux, doutput, cu_seqlen, *, attn_bias_type, attn_
+            max_seqlen,
+            max_seqlen,
+            head_dim,
+            0,    # there is no rng_state in backward currently
```
The backward need to know which element was droped.
How this is done? Maybe extend the comment to specify this?

zlsh80826 replied:
The dropped elements in the softmax_aux tensor are encoded as negative values. This encoding trick is used because the elements after the softmax operation are always greater than 0. By marking the dropped elements as negative values, we can differentiate them from the non-dropped elements. I will enhance the comment here.

zlsh80826 replied:
Comment is enhanced in c8615cf3866fa0266d20739ef221b1ccdb5e45ea

nouiz commented on transformer_engine/jax/csrc/modules.cpp:
```diff
@@ -778,29 +779,40 @@ void SelfFusedAttnMax512Forward(cudaStream_t stream, void **buffers, const char
-    size_t workspace_size =
+    auto workspace_size =
         query_workspace_tensor.shape().data[0] * typeToSize(query_workspace_tensor.dtype());
+    // fused attn workspace size + casted rng state size
```
Why the commented code here and bellow?

zlsh80826 replied:
I will remove them, it should not be here

zlsh80826 replied:
Removed in c8615cf3866fa0266d20739ef221b1ccdb5e45ea

zlsh80826 commented on tests/jax/test_fused_attn.py:
```diff
@@ -322,7 +322,7 @@ def set_input(self, b, s_q, s_kv, h, d, dtype, attn_mask_type, pad_ratio):
         self.scaling_factor = 1. / math.sqrt(d)
         self.dropout_probability = 0.
-        self.dropout_rng = jax.random.PRNGKey(0)
+        self.dropout_rng = jax.random.PRNGKey(0) if self.dropout_probability > 0 else None
```
I can add some dropout sanity test, but the dropout correctness is hard to implement in a unittest since we can't have the bitmatch results.

nouiz replied:
From the description, this PR change the current behavior when fmha is used vs when it is not used.
This isn't good. We need to do something about this.
Could we make both version behave the same?
If not, I think it will need to be an option.

zlsh80826 replied:
`is_training` and `dropout` sanity check are added in f3d21a1e183aee17889ec87dbd8d85854e407847

zlsh80826 replied:
> Could we make both version behave the same?
If not, I think it will need to be an option.

No, we can't make it behave the same. The cuDNN dropout doesn't support the [dropout shape](https://github.com/NVIDIA/TransformerEngine/blob/v0.8.0.1/transformer_engine/jax/flax/transformer.py#L167). I can add an option for it.

nouiz replied:
Why can't we use the cudnn dropout mask shape as the default one?
That way w/wo fmha we will always have the same dropout mask.
The only diff with be the rng used.

zlsh80826 replied:
I changed the default dropout mask shape to be the same as the cudnn's in 4669452f56bab88723e3a8da14a6f427d73509e2. The PR description is also revised, please take a look. Thanks

nouiz reviewed (commented):
I did a quick review.

timmoon10 commented on transformer_engine/common/fused_attn/fused_attn_fp16_bf16_max_seqlen_512.cu:
```diff
@@ -629,27 +640,32 @@ void fused_attn_max_512_fwd_impl(int64_t b, int64_t h, int64_t s_q, int64_t s_kv
-
-        NVTE_CHECK(!enable_dropout,
-                   "dropout probability > 0 in fused_attn_max_512 has not been implemented.");
+        bool enable_dropout = (dropout_probability != 0.0f) && (!is_training);
```
Shouldn't we only apply dropout during training?
```suggestion
        bool enable_dropout = (dropout_probability != 0.0f) && is_training;
```

zlsh80826 replied:
Yes, it was a bug. I fixed it and add the related tests to cover this in 17a183c73c41c988625ae28c779564d3b5fb2afd. Thanks.

ksivaman commented on transformer_engine/common/fused_attn/fused_attn_fp16_bf16_max_seqlen_512.cu:
```diff
@@ -758,16 +776,17 @@ void fused_attn_max_512_fwd_impl(int64_t b, int64_t h, int64_t s_q, int64_t s_kv
+            data_ptrs.insert(std::pair<uint64_t, void *>(S_CONST_ID, &bfloat_cast_scaling_factor));
         } else {
-            std::cerr << "Not supported tensorType." << std::endl;
+            NVTE_ERROR("Not supported tensor type.");
```
```suggestion
            NVTE_ERROR("Unsupported tensor type.");
```

zlsh80826 replied:
Fixed in 17a183c73c41c988625ae28c779564d3b5fb2afd

zlsh80826 commented:
/te-ci

timmoon10 reviewed (approved):
LGTM, pending CI.

zlsh80826 commented:
/te-ci

ptrendx commented:
If I understand correctly, you handle the RNG state needed by cuDNN by just asking the native RNG to give you seed and offset pair every time? To be honest, I do not like this approach. The problem with it is that you have no guarantees how spread out are the numbers generated by the native RNG, so it is possible for different generations to overlap (keep in mind that each call to the fused attention will use multiple values of seed and offset, since each thread gets its own seed and they generate values for different offsets).

What I would prefer instead is having a global (seed, offset) pair, initialized to (prng seed, 0) passed to all of the invocations of fused attention (we should be able to do that since we control the MHA module that would call those functions) and advancing the offset appropriately. The only problem is with seeding - we would need to somehow catch that the user resets the seed in the main PRNG and reset our seed too. How do we know which PRNG a user would like to use?

zlsh80826 commented:
> If I understand correctly, you handle the RNG state needed by cuDNN by just asking the native RNG to give you seed and offset pair every time? To be honest, I do not like this approach. The problem with it is that you have no guarantees how spread out are the numbers generated by the native RNG, so it is possible for different generations to overlap (keep in mind that each call to the fused attention will use multiple values of seed and offset, since each thread gets its own seed and they generate values for different offsets).
> 
> What I would prefer instead is having a global (seed, offset) pair, initialized to (prng seed, 0) passed to all of the invocations of fused attention (we should be able to do that since we control the MHA module that would call those functions) and advancing the offset appropriately. The only problem is with seeding - we would need to somehow catch that the user resets the seed in the main PRNG and reset our seed too. How do we know which PRNG a user would like to use?

The PRNG can be obtained by `prng_impl = config.jax_default_prng_impl`. I understand your preferred approach, it ensures the dropout mask different across different threads/processes. I didn't dig into how the origin JAX handle the prng across threads/process, I just passed the original JAX rng_state fused attention. I will look into it these days.

cyanguwa commented on transformer_engine/common/fused_attn/fused_attn_fp16_bf16_max_seqlen_512.cu:
```diff
@@ -1328,20 +1383,26 @@ void fused_attn_max_512_fwd_kvpacked(size_t batch, size_t q_max_seqlen, size_t k
+    void *devPtrDropoutOffset = nullptr;  // Set according to the rng_state_type
+    if (rng_state_type == DType::kInt32) {
+        devPtrDropoutOffset =
+            static_cast<void *>(static_cast<uint32_t *>(rng_state->data.dptr) + 1);
```
Would it be better to move this conversion to the framework side, so on the C side we always take in a uint64 pair as the seed and offset? The different types and numbers of values in the rng state are caused by the different random number generators used on the framework side.

timmoon10 commented:
Sorry for the late reply. So if I understand correctly, your proposal:
```
seed = PRNG()  # Possibly identical across GPUs
offset = rank * dropout_size
```
I think this would work. Perhaps even better would be to use `seed = rank + PRNG() * num_ranks` or an algorithm like [`boost::hash_combine`](https://boost.sourceforge.net/doc/html/boost/hash_combine.html), since that eliminates the need to know the number of dropout elements.

That said, I think it might be better to leave this outside of TE. If the user is doing any other random operations, they'll still need to initialize the different ranks with different RNG seeds.

zlsh80826 commented:
Thank you for helping review the proposal! It appears that I may not have expressed my proposal clearly. Allow me to provide more detailed information.

My proposal is to generate **different seeds** on different GPUs. Here is the pseudo code for the implementation:

### transformer.py
In each [MultiheadAttention](https://github.com/NVIDIA/TransformerEngine/blob/9573a970b0e30c3e2aae705978a0b7c67b8fa1ee/transformer_engine/jax/flax/transformer.py#L205) module, `n` (the number of GPUs) seeds are generated using jax.random.split:

```
dropout_key = self.make_rng(self.dropout_rng_name) # the original implementation
splitted_dropout_key = jax.random.split(dropout_key, len(jax.devices)) # generate n seeds (n = number of GPUs)
# Then pass splitted_dropout_key to fused_attn (transformer_engine/jax/fused_attn.py)
```
It is important to note that [`self.make_rng`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.make_rng) generates a new key (seed) with each call. As a result, the key (seed) will be different for different MultiheadAttention modules and different iterations.

### transformer_engine/jax/fused_attn.py
To ensure that each GPU has a unique seed, the n seeds are sharded using [`get_fused_sharding_data`](https://github.com/NVIDIA/TransformerEngine/blob/9573a970b0e30c3e2aae705978a0b7c67b8fa1ee/transformer_engine/jax/fused_attn.py#L79).

---

However, I believe your suggestion of `seed = rank + PRNG() * num_ranks` is simpler and ensures that the seeds are different across ranks. We can use this approach instead.

> That said, I wonder if it might be better to leave this outside of TE. If the user is doing any other random operations, they'll still need to initialize the different ranks with different RNG seeds.

As far as I understand, JAX's native random functions (e.g., `jax.random.bernoulli`) use only one seed across GPUs since the tensors are in the global view. This means that the shape of a tensor will be something like `[num_gpus * batch_size, hidden]`. I haven't delved into the underlying implementation, but at the very least, users of native JAX don't need to consider the seed behavior across GPUs.

I believe I can proceed with implementing the details we discussed above so that we can have more concrete code for further discussion. Thanks.

ksivaman commented:
/te-ci

timmoon10 commented on transformer_engine/jax/csrc/utils.h:
```diff
@@ -93,6 +96,27 @@ class cudaDevicePropertiesManager {
     cudaDeviceProp prop_;
 };
 
+class FusedAttnOffsetManager {
```
I don't expect it to cause problems, but it does seem strange to have a global offset value when each attention module may have a different seed.
